### PR TITLE
Add test for generated sparse data to test type detector

### LINKED ISSUE #891: Improve type detection algorithm for spare data
# Overview

See the discussion in the chat:

- https://discordapp.com/invite/Sewv6av
- https://discord.com/channels/695635777199145130/695635777199145133/871416382305239040
- https://discord.com/channels/695635777199145130/695635777199145133/871730478271574047

### PR DESCRIPTION
**Pull request by aivuk**
- fixes #891

Added the tests/detector/test_general.py::test_schema_from_synthetic_sparse_sample that generate a one hundred rows Resource, with a column with different percentage of data from a specific type.

### CODE REVIEW

File: tests/detector/test_general.py [hunk 1]
```diff
@@ -79,6 +79,55 @@ def test_schema_from_sparse_sample():
...
+                    row.append(type_sample[col["type"]]["is"])
+                else:
+                    row.append(type_sample[col["type"]]["not"])
+
+            rows.append(row)
+
+        return rows
+
+    sample = generate_rows(columns=columns)
+    for conf in [0.6, 0.7, 0.8]:
```
Comment (roll) [inline tests/detector/test_general.py]: I think it's better to move it to `@pytest.mark.parametrize`

**Review by roll — COMMENTED** [review]
Great! Just one suggestion

File: tests/detector/test_general.py (hunk 1 above)
Comment (aivuk) [reply tests/detector/test_general.py]: Goot point, done!

**Review by roll — APPROVED** [review]
Great, thanks!

**Comment by aivuk** [comment]
Hi @roll, I did some benchmarks as you requested and I was a bit surprise with how slow the detector is actually. Looks at the visualization created with snakeviz on the cProfile that I created running the attached file:

Screenshot from 2022-05-29 17-20-46

I created a small script to test the performance, a simple modification of the tests for synthetic data but generating a 100.000 rows resource. You can see see the visualization above and investigate yourself with the follow zipped profile file (generated with `python -m cProfile -o p0.prof test-detector-performance.py`):
[schema-detector.prof.zip](https://github.com/frictionlessdata/frictionless-py/files/8793639/schema-detector.prof.zip)

The scripts for the benchmarks are here https://gist.github.com/aivuk/4b95c226b680264f0efae62aef0093da.

The first script test-detector-performance-lines.py I created to use with the module line_profiler (https://github.com/pyutils/line_profiler), but I modified a bit the script to analyze how much time is spend by line in the detector.detect_schema. The results are here:

[performace-detector-lines.txt](https://github.com/frictionlessdata/frictionless-py/files/8793649/performace-detector-lines.txt)
 
I think there is a lot of space for performance improvement in the schema detector and I'm doing some additional experiments and I will keep you update.

**Comment by aivuk** [comment]
@roll Investigating a bit further raised some questions. I created a csv with 100.000 rows and tried to infer the schema in two ways:

In [1]: from frictionless import Resource
```

In [2]: %time r = Resource('t.csv') ; r.infer() ; r.schema
CPU times: user 71 ms, sys: 7.97 ms, total: 79 ms
Wall time: 78.5 ms
Out[2]: 
{'fields': [{'name': '1', 'type': 'integer'},
            {'name': '3.14', 'type': 'number'},
            {'name': 'string', 'type': 'string'}]}


```

And the other way I created the same csv content in a Python list called `rs` and used the detector as:

```
In [10]: %time d.detect_schema(rs)
CPU times: user 2.16 s, sys: 4.56 ms, total: 2.17 s
Wall time: 2.17 s
Out[10]: 
{'fields': [{'name': 'field1', 'type': 'integer'},
            {'name': 'field2', 'type': 'number'},
            {'name': 'field3', 'type': 'string'}]}

```

305 slower than with .infer(). Any idea why? I didn't quiet get yet how infer() is doing the magic.

**Comment by aivuk** [comment]
Using .infer() is not much slower than loading the one thousand rows csv directly from pandas:

```
In [1]: import pandas as pd

In [2]: %time d = pd.read_csv('t.csv')
CPU times: user 16.9 ms, sys: 4.82 ms, total: 21.7 ms
Wall time: 21.2 ms

In [3]: from frictionless import Resource

In [4]: %time r = Resource('t.csv') ; r.infer()
CPU times: user 57 ms, sys: 3.94 ms, total: 60.9 ms
Wall time: 60.1 ms


```

**Pull request merged**